src/WidgetBase.ts

```diff
@@ -169,6 +175,14 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> implement
 		// Do nothing by default.
 	}
 
+	protected onAttach(): void {
```

Comment:
Any thought to this just being an optional method on `WidgetBaseInterface`?

Reply:
also, you will essentially always do a noop on _most_ widgets on attachment, and I don't know the performance impact, but invoking potentially hundreds of the same function seems 😕 where as testing for presence and only calling if present feels like it would be more efficient...

Reply:
We do the same for the `onElementCreated` / `onElementUpdated`, these guys get called on `requestIdleCallback` so certainly should reduce the performance impact - but I can look at the optional method.

Reply:
Problem is I don't want to put it on the interface (it's protected), so I don't think we can type is as such.

Reply:
Yeah, optional abstract isn't allowed and protected cannot be enforced on an interface... You ruin all my best ideas... 😝